psi4/src/psi4/libfock/CompositeJK.cc

```diff
@@ -144,186 +145,273 @@ Matrix compute_esp_bound(const BasisSet &primary) {
+            {"DFT_BLOCK_MIN_POINTS", 100},
+            {"DFT_BLOCK_MAX_POINTS", 256},
+        };
+        std::map<std::string, double> grid_final_float_options = {
+            {"DFT_BASIS_TOLERANCE",   options_.get_double("COSX_BASIS_TOLERANCE")},
+            {"DFT_BS_RADIUS_ALPHA",   1.0},
+            {"DFT_PRUNING_ALPHA",     1.0},
+            {"DFT_BLOCK_MAX_RADIUS",  3.0},
+            {"DFT_WEIGHTS_TOLERANCE", 1e-15},
+        };
+        grid_final_ = std::make_shared<DFTGrid>(primary_->molecule(), primary_, grid_final_int_options, grid_final_str_options, grid_final_float_options, options_);
+
+        timer_off("CompositeJK: COSX Grid Construction");
+
+        // => Overlap Fitting Metric <= //
+
+        // Fit an overlap metric (Q) for both grids to reduce numerical error
+
+        // DOI 10.1063/1.3646921, EQ. 18
+        // Note: the above reference defines Q as S_an @ S_num^{-1} @ X
+        // Here, Q refers to just S_ @ S_num^{-1} (no X)
+        // This Q is contracted with X later to agree with the literature definition
 
-    // solve: Q_final_ = S_an @ S_num_final_^{-1}
-    Q_final_ = S_an->clone();
-    C_DGESV(nbf, nbf, S_num_final.pointer()[0], nbf, ipiv.data(), Q_final_->pointer()[0], nbf);
+        timer_on("CompositeJK: COSX Numeric Overlap");
 
-    timer_off("Overlap Metric Solve");
+        // compute the numeric overlap matrix for each grid
+        auto S_num_init = compute_numeric_overlap(*grid_init_, primary_);
+        auto S_num_final = compute_numeric_overlap(*grid_final_, primary_ );
 
+        timer_off("CompositeJK: COSX Numeric Overlap");
+
+        timer_on("CompositeJK: COSX Analytic Overlap");
+
+        // compute the analytic overlap matrix
+        MintsHelper helper(primary_, options_);
+        auto S_an = helper.ao_overlap();
+
+        timer_off("CompositeJK: COSX Analytic Overlap");
+
+        // form the overlap metric (Q) for each grid
+
+        timer_on("CompositeJK: COSX Overlap Metric Solve");
+
+        int nbf = primary_->nbf();
+        std::vector<int> ipiv(nbf);
+
+        // solve: Q_init_ = S_an @ S_num_init_^{-1}
+        Q_init_ = S_an->clone();
+        C_DGESV(nbf, nbf, S_num_init.pointer()[0], nbf, ipiv.data(), Q_init_->pointer()[0], nbf);
+
+        // solve: Q_final_ = S_an @ S_num_final_^{-1}
+        Q_final_ = S_an->clone();
+        C_DGESV(nbf, nbf, S_num_final.pointer()[0], nbf, ipiv.data(), Q_final_->pointer()[0], nbf);
+
+        timer_off("CompositeJK: COSX Overlap Metric Solve");
+    } else {
+        throw PSIEXCEPTION("Invalid Composite K algorithm selected!");
+    }
+}
+
+size_t CompositeJK::num_computed_shells() {
+    //no bench data returned - to come in a future update
+    return JK::num_computed_shells();
 }
 
-size_t DFJCOSK::memory_estimate() {
+size_t CompositeJK::memory_estimate() {
     return 0;  // Memory is O(N^2), which psi4 counts as effectively 0
 }
 
-void DFJCOSK::print_header() const {
+void CompositeJK::print_header() const {
```

Comment:
The header printing scheme was changed a little bit for CompositeJK - this is worth looking at.